Ingress/controllers/nginx-third-party/controller.go

```diff
@@ -0,0 +1,237 @@
+		third party resource that contains the custom configuration to be used in nginx.
+		Takes the form namespace/name.`)
+
+	healthzPort = flags.Int("healthz-port", healthPort, "port for healthz endpoint.")
+)
+
+type k8sService struct {
+	ServiceName string
+	ServicePort string
+	Namespace   string
+	ExposedPort string
+}
+
+func registerHandlers() {
+	http.HandleFunc("/healthz", func(w http.ResponseWriter, r *http.Request) {
+		w.WriteHeader(200)
+		w.Write([]byte("ok"))
+	})
+
+	glog.Fatalf(fmt.Sprintf("%v", http.ListenAndServe(fmt.Sprintf(":%v", *healthzPort), nil)))
+}
+
+func main() {
+	pflagGlog(flags)
+
+	flags.Parse(os.Args)
+
+	if *defaultSvc == "" {
+		glog.Fatalf("Please specify --default-backend")
+	}
+
+	var ingClient unversioned.IngressInterface
+
+	kubeClient, err := unversioned.NewInCluster()
+	if err != nil {
+		glog.Fatalf("failed to create client: %v", err)
+	} else {
+		ingClient = kubeClient.Extensions().Ingress(api.NamespaceAll)
+	}
+
+	funcMap := template.FuncMap{
+		"getSSLHost": getSSLHost,
+		"empty": func(input interface{}) bool {
+			check, ok := input.(string)
+			if ok {
+				return len(check) == 0
+			}
+
+			return true
+		},
+		"dict": func(values ...interface{}) (map[string]interface{}, error) {
+			if len(values)%2 != 0 {
+				return nil, errors.New("invalid dict call")
+			}
+			dict := make(map[string]interface{}, len(values)/2)
+			for i := 0; i < len(values); i += 2 {
+				key, ok := values[i].(string)
+				if !ok {
+					return nil, errors.New("dict keys must be strings")
+				}
+				dict[key] = values[i+1]
+			}
+			return dict, nil
+		},
+	}
+
+	tmpl, _ := template.New("nginx.tmpl").Funcs(funcMap).ParseFiles("./nginx.tmpl")
+	rateLimiter := util.NewTokenBucketRateLimiter(0.1, 1)
+
+	known := &extensions.IngressList{}
+	defCfg := defNginxCfgParams()
+	runningCfg := make(map[string]interface{})
+
+	sslCertificates := createSSLCerts(sslDirectory)
+
+	servicesL4 := getTcpServices(kubeClient, *tcpServices)
+	if len(servicesL4) == 0 {
+		glog.Info("no tcp service defined. skipping")
+	} else {
+		glog.Infof("tcp service/s defined: %v", servicesL4)
+	}
+
+	defBackendSvc := getService(kubeClient, *defaultSvc)
+	defErrorSvc := getService(kubeClient, *customErrorSvc)
+	defDhFile := dhParamFile(sslDirectory)
+
+	// we need to configure the nginx resolver parameter
+	// http://nginx.org/en/docs/http/ngx_http_core_module.html#resolver
+	// if kube-dns is not working it will return a fatal error.
+	defNgxResolver := skydnsIPPort(kubeClient)
+
+	// we need to start nginx and terminate the process if it dies
+	go startNginx()
+
+	conf := make(map[string]interface{})
+	conf["sslCertificates"] = sslCertificates
+	conf["tcpServices"] = servicesL4
+	conf["defErrorSvc"] = defErrorSvc
+	conf["defBackend"] = defBackendSvc
+	conf["sslDHParam"] = defDhFile
+
+	go registerHandlers()
+
+	// Controller loop
+	for {
+		rateLimiter.Accept()
```

Comment:
Can we make this control loop more generic? something like:
foo := newTemplatedNginx(nginx.tmpl)
foo.getConfig
foo.modifyConfig(options...)
foo.writeConfig

It would be *really* cool if we could make an interface and combine this control loop with the one in nginx-toy, but I'll not block on that